samples/_emscripten/_audio/EmscriptenAudio/Voice.cpp

```diff
@@ -0,0 +1,345 @@
+{
+	const auto &bus = mBusses.at( busId );
+
+	if( bus.mPan )
+		return bus.mPan;
+	else
+		return bus.mGain;
+}
+
+// ----------------------------------------------------------------------------------------------------
+// Voice
+// ----------------------------------------------------------------------------------------------------
+
+VoiceRef Voice::create( const CallbackProcessorFn &callbackFn, const Options &options )
+{
+	VoiceRef result( new VoiceCallbackProcessor( callbackFn, options ) );
+	MixerImpl::get()->addVoice( result, options );
+
+	return result;
+}
+
+bool Voice::isDoneLoading(){
+	return mDoneParsing;
+}
+
+void Voice::create( const SourceFileRef &sourceFile, std::function<void(VoiceSamplePlayerNodeRef)> cb, const Options &options){
+	new VoiceSamplePlayerNode( sourceFile, options , cb );
+}
+VoiceSamplePlayerNodeRef Voice::create( const SourceFileRef &sourceFile, const Options &options )
+{
+	VoiceSamplePlayerNodeRef result( new VoiceSamplePlayerNode( sourceFile, options ) );
+	MixerImpl::get()->addVoice( result, options );
+
+	return result;
+}
+
+Voice::~Voice()
+{
+	MixerImpl::get()->removeVoice( mBusId );
+}
+
+void Voice::clearBufferCache()
+{
+	MixerImpl::get()->clearBufferCache();
+}
+
+float Voice::getVolume() const
+{
+	return MixerImpl::get()->getBusVolume( mBusId );
+}
+
+float Voice::getPan() const
+{
+	return MixerImpl::get()->getBusPan( mBusId );
+}
+
+void Voice::setVolume( float volume )
+{
+	MixerImpl::get()->setBusVolume( mBusId, volume );
+}
+
+void Voice::setPan( float pan )
+{
+	MixerImpl::get()->setBusPan( mBusId, pan );
+}
+
+void Voice::start()
+{
+	getInputNode()->enable();
+}
+
+void Voice::pause()
+{
+	getInputNode()->disable();
+}
+
+void Voice::stop()
+{
+	getInputNode()->disable();
+}
+
+bool Voice::isPlaying() const
+{
+	return getInputNode()->isEnabled();
+}
+
+NodeRef Voice::getOutputNode() const
+{
+	return MixerImpl::get()->getOutputNode( mBusId );
+}
+
+// ----------------------------------------------------------------------------------------------------
+// VoiceSamplePlayerNode
+// ----------------------------------------------------------------------------------------------------
+
+
+VoiceSamplePlayerNode::VoiceSamplePlayerNode( const SourceFileRef &sourceFile, const Options &options, std::function<void(VoiceSamplePlayerNodeRef)> cb )
+	: Voice()
+{
+
+
+	#ifndef CINDER_EMSCRIPTEN
+		size_t requiredSampleRate = audio::master()->getSampleRate();
+		SourceFileRef sf = requiredSampleRate == sourceFile->getSampleRate() ? sourceFile : sourceFile->cloneWithSampleRate( requiredSampleRate );
+
+		if( sf->getNumFrames() <= options.getMaxFramesForBufferPlayback() ) 
+		{
+			BufferRef buffer = MixerImpl::get()->loadBuffer( sf );
+			mNode = Context::master()->makeNode( new BufferPlayerNode( buffer ) );
+		} 
+		else {
+			mNode = Context::master()->makeNode( new FilePlayerNode( sf ) );
+		}
+	#else
+
+
+		auto loader = (WebAudioLoader*)sourceFile.get();
+
+		loader->init( [=]()->void
```

Comment:
You usually want to avoid copying a local variables into a lambda (the '=' in `[=]`), instead it's best to pass in the `this` pointer and anything else explicitly. Also, I'm not sure if the `WebAudioLoader *loader` will be valid if this callback is async. By the way, is it (`WebAudioLoader::init()`) async? If it isn't what is the intention of using a callback?

Reply:
👍 Did not know that about lambdas, changed things up a bit.

`WebAudioLoader::init()` is indeed async. The callback you pass to it is called once the audio is loaded on the Web side of things. The path essentially is 

* load audio on JS side
* local lambda in `WebAudioLoader::init()` is then called to parse the information into a local `BufferRef` 
* finally lambda passed to `WebAudioLoader::init()` (note - now `SourceFileWebAudio`) is called to finish loading audio.

`loader` does still work from what I can see